python/ray/tests/test_basic_5.py

```diff
@@ -119,5 +121,95 @@ def test_internal_kv(ray_start_regular):
         kv._internal_kv_list("@namespace_abc", namespace="n")
 
 
+def get_gcs_memory_used():
+    m = sum([
+        process.memory_info().rss for process in psutil.process_iter()
+        if process.name() in ("gcs_server", "redis-server")
+    ])
+    print(m)
+    return m
+
+
+def function_entry_num(job_id):
+    from ray.ray_constants import KV_NAMESPACE_FUNCTION_TABLE
+    return len(_internal_kv_list(b"IsolatedExports:" + job_id,
+                                 namespace=KV_NAMESPACE_FUNCTION_TABLE)) + \
+        len(_internal_kv_list(b"RemoteFunction:" + job_id,
+                              namespace=KV_NAMESPACE_FUNCTION_TABLE)) + \
+        len(_internal_kv_list(b"ActorClass:" + job_id,
+                              namespace=KV_NAMESPACE_FUNCTION_TABLE))
+
+
+@pytest.mark.skipif(
+    client_test_enabled(),
+    reason="client api doesn't support namespace right now.")
+def test_function_table_gc(call_ray_start):
+    """This test tries to verify that function table is cleaned up
+    after job exits.
+    """
+
+    def f():
+        data = "0" * 1024 * 1024  # 1MB
+
+        @ray.remote
+        def r():
+            nonlocal data
+
+            @ray.remote
+            class Actor:
+                pass
+
+        return r.remote()
+
+    ray.init(address="auto", namespace="b")
+
+    # It should use > 500MB data
+    ray.get([f() for _ in range(500)])
+
+    assert get_gcs_memory_used() > 500 * 1024 * 1024
+    job_id = ray.worker.global_worker.current_job_id.binary()
+
+    ray.shutdown()
+
+    # now check the function table is cleaned up after job finished
+    ray.init(address="auto", namespace="a")
+    wait_for_condition(lambda: function_entry_num(job_id) == 0)
```

Comment:
Could we also check memory <500 at this point?

Reply:
@ericl  I used to check this, but for memory kv, it won't work since the eviction of virtual memory is decided by the os.

For 500MB, in redis, it'll drop, but in memory kv, it'll still be there. But if it's like GBs data and job exited, it'll drop to hundreds MB.

Reply:
I see, maybe we can check the total redis memory instead then, but it's OK.

Reply:
I can do this, if it's redis as storage backend, check it. If it's memory backend, just skip the check.